reactor-netty-http/src/main/java/reactor/netty/http/server/AbstractHttpServerMetricsHandler.java

```diff
@@ -83,94 +88,112 @@ public void channelInactive(ChannelHandlerContext ctx) {
+			dataSent += extractProcessedDataFromBuffer(msg);
+
+			if (msg instanceof LastHttpContent) {
+				promise.addListener(future -> {
+					try {
+						withHttpServerOperations(ctx, ops -> {
+							recordWrite(ops, uriTagValue == null ? ops.path : uriTagValue.apply(ops.path),
+									ops.method().name(), ops.status().codeAsText().toString());
+							if (!ops.isHttp2() && ops.hostAddress() != null) {
+								// This metric is not applicable for HTTP/2
+								// ops.hostAddress() == null when request decoding failed, in this case
+								// we do not report active connection, so we do not report inactive connection
+								recordInactiveConnection(ops);
+							}
+						});
+					} catch (RuntimeException e) {
+						log.warn("Exception caught while recording metrics.", e);
+						// Allow request-response exchange to continue, unaffected by metrics problem
 					}
-				}
 
-				dataSent = 0;
-			});
+					dataSent = 0;
+				});
+			}
+		} catch (RuntimeException e) {
+			log.warn("Exception caught while recording metrics.", e);
+			// Allow request-response exchange to continue, unaffected by metrics problem
+		} finally {
+			//"FutureReturnValueIgnored" this is deliberate
+			ctx.write(msg, promise);
 		}
-
-		//"FutureReturnValueIgnored" this is deliberate
-		ctx.write(msg, promise);
 	}
 
 	@Override
 	public void channelRead(ChannelHandlerContext ctx, Object msg) {
-		if (msg instanceof HttpRequest) {
-			dataReceivedTime = System.nanoTime();
-			ChannelOperations<?, ?> channelOps = ChannelOperations.get(ctx.channel());
-			if (channelOps instanceof HttpServerOperations) {
-				HttpServerOperations ops = (HttpServerOperations) channelOps;
-				if (!ops.isHttp2()) {
-					// This metric is not applicable for HTTP/2
-					recordActiveConnection(ops);
-				}
+		try {
+			if (msg instanceof HttpRequest) {
+				dataReceivedTime = System.nanoTime();
+				withHttpServerOperations(ctx, ops -> {
+					if (!ops.isHttp2()) {
+						// This metric is not applicable for HTTP/2
+						recordActiveConnection(ops);
+					}
+				});
 			}
-		}
 
-		if (msg instanceof ByteBufHolder) {
-			dataReceived += ((ByteBufHolder) msg).content().readableBytes();
-		}
-		else if (msg instanceof ByteBuf) {
-			dataReceived += ((ByteBuf) msg).readableBytes();
-		}
+			dataReceived += extractProcessedDataFromBuffer(msg);
 
-		if (msg instanceof LastHttpContent) {
-			ChannelOperations<?, ?> channelOps = ChannelOperations.get(ctx.channel());
-			if (channelOps instanceof HttpServerOperations) {
-				HttpServerOperations ops = (HttpServerOperations) channelOps;
-				recordRead(ops, uriTagValue == null ? ops.path : uriTagValue.apply(ops.path), ops.method().name());
-			}
+			if (msg instanceof LastHttpContent) {
+				withHttpServerOperations(ctx, ops -> {
+					recordRead(ops, uriTagValue == null ? ops.path : uriTagValue.apply(ops.path), ops.method().name());
+				});
 
-			dataReceived = 0;
+				dataReceived = 0;
+			}
+		} catch (RuntimeException e) {
```

Comment:
Do we need this `try/catch` for the whole block or it is more appropriate to have it only for the recording? Is it possible to have a `RuntimeException` not related to metrics?

Reply:
Similar to the other handler. If it's save to assume the other operations will not throw I'd gladly move the try/catch around the actual record operations.